Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/business/wrapper/Mappings.kt

```diff
@@ -140,9 +141,39 @@ internal fun List<DccValidationRule>.filterRelevantRules(
             rule.certificateType.uppercase() == certificateType.uppercase()
     }.filter { rule ->
         rule.validFromInstant <= validationClock && rule.validToInstant >= validationClock
+    }.groupBy { it.identifier }.mapNotNull { it.value.takeHighestVersion() }
+}
+
+internal fun List<DccValidationRule>.takeHighestVersion(): DccValidationRule? {
+    return maxWithOrNull(comparator)
+}
+
+private val comparator = object : Comparator<DccValidationRule> {
+    override fun compare(o1: DccValidationRule, o2: DccValidationRule): Int {
+        val version1 = o1.version.parseVersion()
+        val version2 = o2.version.parseVersion()
+        (0 until min(version1.size, version2.size)).forEach {
+            try {
+                val diff = Integer.parseInt(version1[it]) - Integer.parseInt(version2[it])
+                if (diff != 0) return diff
+            } catch (e: Exception) {
+                // compare as String
+                val word1 = version1[it]
+                val word2 = version2[it]
+                if (word1 < word2) {
+                    return -1
+                }
+                if (word1 > word2) {
+                    return 1
+                }
+            }
+        }
+        return version1.size - version2.size
     }
 }
 
+internal fun String.parseVersion(): List<String> = this.split('.')
```

Comment:
What do we do if the version scheme is incorrect?
Should we enforce that?
e.g. 
```kotlin
Regex("(\d+)\.(\d+)\.(\d+))")
```
if it does not match, assume version is `0` lower then all others?

/cc @mlenkeit

Reply:
It's enforced on the server by schema validation; should never happen ™️. I would suggest to handle this case in some way that does not crash the client and log an error message.

Reply:
Version `0` as fallback in case parsing fails seems like a good idea